// safe-js/dist/dns.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.listServices = exports.listLongNames = exports.createLongName = exports.addService = exports.manifest = undefined;

var _isomorphicFetch = require('isomorphic-fetch');

var _isomorphicFetch2 = _interopRequireDefault(_isomorphicFetch);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
* Manifest for Beaker: 
* https://github.com/pfrazee/beaker/blob/master/doc/authoring-plugins.md#api-manifests
*/
var manifest = exports.manifest = {
  addService: 'promise',
  createLongName: 'promise',
  listLongNames: 'promise',
  listServices: 'promise'
};

// add service
var addService = exports.addService = function addService(token, longName, serviceName, serviceHomeDirPath, isPathShared) {
  var url = _utils.SERVER + 'dns';
  var rootPath = isPathShared ? _utils.ROOT_PATH.DRIVE : _utils.ROOT_PATH.APP;

  var payload = {
    method: 'PUT',
    headers: {
      Authorization: 'Bearer ' + token,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify({
      longName: longName,
      serviceName: serviceName,
      rootPath: rootPath,
      serviceHomeDirPath: serviceHomeDirPath
    })
  };

  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    return (0, _utils.parseResponse)(response);
  });
};

var createLongName = exports.createLongName = function createLongName(token, longName) {
  var url = _utils.SERVER + 'dns/' + longName;
  var payload = {
    method: 'POST',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    return (0, _utils.parseResponse)(response);
  });
};

// get dns list
var listLongNames = exports.listLongNames = function listLongNames(token) {
  var url = _utils.SERVER + 'dns';
  var payload = {
    method: 'GET',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };

  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    return (0, _utils.parseResponse)(response);
  });
};

// get service
var listServices = exports.listServices = function listServices(token, longName) {
  var url = _utils.SERVER + 'dns/' + longName;
  var payload = {
    method: 'GET',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    return (0, _utils.parseResponse)(response);
  });
};
