// safe-js/dist/structured_data.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.makeUnclaimable = exports.getMetadata = exports.del = exports.deserialise = exports.serialise = exports.dropHandle = exports.readData = exports.post = exports.put = exports.getDataIdHandle = exports.isSizeValid = exports.getHandle = exports.updateData = exports.create = exports.manifest = undefined;

var _crypto = require('crypto');

var _crypto2 = _interopRequireDefault(_crypto);

var _isomorphicFetch = require('isomorphic-fetch');

var _isomorphicFetch2 = _interopRequireDefault(_isomorphicFetch);

var _utils = require('./utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SD_ENDPOINT = _utils.SERVER + 'structured-data/';

var manifest = exports.manifest = {
  create: 'promise',
  del: 'promise',
  dropHandle: 'promise',
  deserialise: 'promise',
  getDataIdHandle: 'promise',
  getHandle: 'promise',
  getMetadata: 'promise',
  put: 'promise',
  post: 'promise',
  readData: 'promise',
  serialise: 'promise',
  updateData: 'promise',
  makeUnclaimable: 'promise',
  isSizeValid: 'promise'
};

/**
 *
 * @param token
 * @param name
 * @param typeTag
 * @param data- as base64 string
 * @param cipherOptsHandle
 */
var create = exports.create = function create(token, name) {
  var typeTag = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : _utils.SD_DEFAULT_TYPE_TAG;
  var data = arguments[3];
  var cipherOptsHandle = arguments[4];

  if (typeof name === 'string') {
    name = _crypto2.default.createHash('sha256').update(name).digest('base64');
  }
  var body = {
    name: name,
    typeTag: typeTag,
    cipherOpts: cipherOptsHandle,
    data: data
  };
  var payload = {
    method: 'POST',
    headers: {
      Authorization: 'Bearer ' + token,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(body)
  };
  return (0, _isomorphicFetch2.default)(SD_ENDPOINT, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'StructuredData creation failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: SD_ENDPOINT
      });
    }
    return (0, _utils.parseResponse)(response);
  });
};

var updateData = exports.updateData = function updateData(token, handleId, data, cipherOptsHandle) {
  var body = {
    data: data,
    cipherOpts: cipherOptsHandle
  };
  var payload = {
    method: 'PATCH',
    headers: {
      Authorization: 'Bearer ' + token,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(body)
  };
  var url = SD_ENDPOINT + handleId;
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Update StructuredData failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return response;
  });
};

var getHandle = exports.getHandle = function getHandle(token, dataIdHandle) {
  var payload = {
    method: 'GET'
  };
  if (token) {
    payload.headers = {
      Authorization: 'Bearer ' + token
    };
  }
  var url = SD_ENDPOINT + 'handle/' + dataIdHandle;
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error(JSON.stringify({ error: 'Get StructuredData handle failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      }));
    }
    return (0, _utils.parseResponse)(response);
  });
};

var isSizeValid = exports.isSizeValid = function isSizeValid(token, handleId) {
  var payload = {
    method: 'GET'
  };
  if (token) {
    payload.headers = {
      Authorization: 'Bearer ' + token
    };
  }
  var url = SD_ENDPOINT + 'validate-size/' + handleId;
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Validating StructuredData handle failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return (0, _utils.parseResponse)(response);
  });
};

var getDataIdHandle = exports.getDataIdHandle = function getDataIdHandle(token, handleId) {
  var payload = {
    method: 'GET'
  };
  if (token) {
    payload.headers = {
      Authorization: 'Bearer ' + token
    };
  }
  var url = SD_ENDPOINT + 'data-id/' + handleId;
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Get DataId handle of StructuredData failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return (0, _utils.parseResponse)(response);
  });
};

var put = exports.put = function put(token, handleId) {
  var payload = {
    method: 'PUT',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };
  var url = SD_ENDPOINT + handleId;
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'PUT of StructuredData failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return response;
  });
};

var post = exports.post = function post(token, handleId) {
  var payload = {
    method: 'POST',
    headers: {
      Authorization: 'Bearer ' + token
    }
  };
  var url = SD_ENDPOINT + handleId;
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'PUT of StructuredData failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return response;
  });
};

var readData = exports.readData = function readData(token, handleId, version) {
  var url = '' + SD_ENDPOINT + handleId + '/' + version;
  var payload = {
    method: 'GET'
  };
  if (token) {
    payload.headers = {
      'Authorization': 'Bearer ' + token
    };
  }
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error('Read StructuredData failed with status ' + response.status + ' ' + response.statusText);
    }
    return response.buffer();
  });
};

var dropHandle = exports.dropHandle = function dropHandle(token, handleId) {
  var payload = {
    method: 'DELETE'
  };
  if (token) {
    payload.headers = {
      'Authorization': 'Bearer ' + token
    };
  }
  var url = SD_ENDPOINT + 'handle/' + handleId;
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Drop StructuredData handle failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return response;
  });
};

var serialise = exports.serialise = function serialise(token, handleId) {
  var url = SD_ENDPOINT + 'serialise/' + handleId;
  var payload = {
    method: 'GET'
  };
  if (token) {
    payload.headers = {
      'Authorization': 'Bearer ' + token
    };
  }
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Serialise StructuredData handle failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return response.buffer();
  });
};

var deserialise = exports.deserialise = function deserialise(token, data) {
  var url = SD_ENDPOINT + 'deserialise';
  var payload = {
    method: 'POST',
    body: data
  };
  if (token) {
    payload.headers = {
      'Authorization': 'Bearer ' + token
    };
  }
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Deserialise StructuredData handle failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return (0, _utils.parseResponse)(response);
  });
};

var del = exports.del = function del(token, handleId) {
  var url = '' + SD_ENDPOINT + handleId;
  var payload = {
    method: 'DELETE'
  };
  if (token) {
    payload.headers = {
      'Authorization': 'Bearer ' + token
    };
  }
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Delete StructuredData failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return response;
  });
};

var getMetadata = exports.getMetadata = function getMetadata(token, handleId) {
  var url = SD_ENDPOINT + 'metadata/' + handleId;
  var payload = {
    method: 'GET'
  };
  if (token) {
    payload.headers = {
      'Authorization': 'Bearer ' + token
    };
  }
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Get StructuredData meta data failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return (0, _utils.parseResponse)(response);
  });
};

var makeUnclaimable = exports.makeUnclaimable = function makeUnclaimable(token, handleId) {
  var url = SD_ENDPOINT + 'unclaim/' + handleId;
  var payload = {
    method: 'DELETE'
  };
  if (token) {
    payload.headers = {
      'Authorization': 'Bearer ' + token
    };
  }
  return (0, _isomorphicFetch2.default)(url, payload).then(function (response) {
    if (response.status !== 200) {
      throw new Error({ error: 'Make StructuredData Unclaimable failed with status ' + response.status + ' ' + response.statusText,
        errorPayload: payload,
        errorUrl: url
      });
    }
    return response;
  });
};
