// safe-js/dist/utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
var VERSION = exports.VERSION = '0.5';
var SERVER = exports.SERVER = 'http://localhost:8100/' + VERSION + '/';
var SD_DEFAULT_TYPE_TAG = exports.SD_DEFAULT_TYPE_TAG = 500; // UNVERSIONED
var ROOT_PATH = exports.ROOT_PATH = {
    APP: 'app',
    DRIVE: 'drive'
};

var failParsing = function failParsing(response) {
    //handle unauthorised requests
    if (response.status === 401 || response.status === 404) {
        return Promise.reject(response);
    }

    return response.clone().json().then(function (json) {
        return Promise.reject(new Error(json));
    });
};

var parseResponse = exports.parseResponse = function parseResponse(response) {
    if (response.status !== 200) {
        return failParsing(response);
    } else {
        return response.json();
    }
};

var checkBooleanResponse = exports.checkBooleanResponse = function checkBooleanResponse(response) {
    if (response.status !== 200) {
        return failParsing(response);
    } else {
        return true;
    }
};
